### DISCUSSION TITLE
조건검색 관련

### DISCUSSION THREAD

**Discussion by fckorea** [elbakramer/koapy#7] [Q&A]
안녕하세요~
이번에 koapy를 이용해서 개발을 하려는 사람입니다.
다름이 아니라, 개발하는 도중에 아래와 같은 이슈가 있어 글 남겨 드립니다.

* examples에 있는 condition.py를 실행하였을때
  * 첫번째는 잘 동작함
  * 두번째 실행 시 에러 발생 ("{"created":"@1612246211.692000000","description":"Error received from peer ipv4:127.0.0.1:5943","file":"src/core/lib/surface/call.cc","file_line":1063,"grpc_message":"Exception iterating responses: Failed to send condition","grpc_status":2}")
  * 이후 계속 에러 발생

잘 동작할 당시에 condition.py 종료 전에 트레이아이콘에 있는 프로그램을 Exit로 종료를 하면 두번째 실행시에도 잘 됩니다. (매번 Exit로 종료하면 잘됩니다)
gRPC에서 정상적으로 종료가 이뤄지지 않은 부분이 있지 않을까 싶은데, 제가 해결하지 못하는 부분이라 글 남겨드립니다~!

다른 패키지를 써봐도 KOAPY가 제일 잘 만들어진거 같아 꼭 사용하고 싶습니다ㅎㅎㅎ
감사합니다~!

**Comment by elbakramer** [comment]
안녕하세요.

붙여넣어주신 에러 내용을 봤을 때 에러가 발생한 위치는 아래인거 같구요.

https://github.com/elbakramer/koapy/blob/6218519cbd950998a2d5d5973bde8f9f4df9d858/koapy/grpc/event/KiwoomOpenApiEventHandlers.py#L1016-L1019

설명해주신 동작을 살펴봤을때 아마도 키움쪽에서 조건검색을 제한하는 부분으로 인해서 에러가 발생하는 듯 합니다.
프로그램을 종료하고 다시 실행하는 경우 제대로 동작하는건 해당 제한이 리셋되면서 그런거 같네요.

조건검색 제한과 관련해서 KOAStudio 에 설명되어 있는 내용을 그대로 가져오면 아래와 같습니다.

```
[조건검색 제한]
조건검색(실시간 조건검색 포함)은 시세조회와 관심종목조회와 합산해서 1초에 5회만 요청 가능하며 1분에 1회로 조건검색 제한됩니다.
조건검색 제한에 대한 자세한 내용은 하단을 참고해 주세요.

10개 조건검색식을 한번에 모두 조회하는 프로그램이 있으며 조건검색만 요청한다고 가정해서 설명하면 다음과 같습니다.

------------------------------------------------------------------------------------------------------------------------------------

첫번째 제한조건 : 1초에 5회만 조회가능
두번째 제한조건 : 조건별 1분당 1회로 제한(실시간 조건검색 수신에는 영향없음)

09:00:00  조회 시작
1번부터 5번 조건식은 조회성공(첫번째 제한조건, 두번째 제한조건 모두 만족)
6번부터 10번 조건식은 조회실패(첫번째 제한조건)

09:00:01 조회 재시작(1초후 재조회)
1번부터 5번 조건식은 조회실패(두번째 제한조건)
6번부터 10번 조건식은 조회성공(첫번째 제한조건, 두번째 제한조건 모두 만족)
... ... ...

09:01:00 조회시작 (첫조회 1분후)
1번부터 5번 조건식은 조회성공(첫번째 제한조건, 두번째 제한조건 모두 만족)
6번부터 10번 조건식은 조회실패(첫번째 제한조건)

09:00:01 조회 재시작(1분 1초후 재조회)
1번부터 5번 조건식은 조회실패(두번째 제한조건)
6번부터 10번 조건식은 조회성공(첫번째 제한조건, 두번째 제한조건 모두 만족)

------------------------------------------------------------------------------------------------------------------------------------
```

위의 제한조건보다 자주 조회가 필요한 경우라면 1회성으로 조회하는 기능 대신에 실시간 조건검색 기능을 사용하시는 쪽을 권장드립니다.

이외에 관련해서 자세한 내용은 아래 페이지에서 KOAStudio 를 받으시고, 개발가이드의 조건검색 관련 내용을 확인해 보시기 바랍니다.
https://www3.kiwoom.com/nkw.templateFrameSet.do?m=m1408010600

**Reply by solverkim** [reply]
좋은 프로그램 Open해 주셔서 감사 드립니다.
저도 fckorea님과 동일한 증상이 나타나고 있어서 방법을 찾아 보고 있는데.. 어렵네요
tray에서 Exit해도 안되고 pyqt5로 별도로 작성해 놓았던 프로그램으로 한번 접속하고 나면
조건 검색이 1회 제대로 수행됩니다.
다음에는 수행되지 않구요.
다시 pyqt로 사용하던 프로그램으로 한번 접속하면 condition.py에서 조건 검색이 1회 실행 가능합니다.
기존에 만들어 놓았던 pyqt5는 위에 말씀하신 kiwoom 조건검색 제한에만 해당되지 않으면 정상조회가 되는데..
koapy에서는 이상하게 조회가 되지 않네요.

문제의 원인을 점검할 수 있는 방법이 있을까요?

**Reply by elbakramer** [reply]
안녕하세요. @solverkim 

조건검색 관련해서 말씀주신 동작을 좀 살펴봤는데요.
조건정보를 불러오는 `EnsureConditionLoaded()` 함수 내부의 조건판단 부분과 관련해서 제가 코드를 잘못짜서..
해당 함수를 호출하더라도 제대로 조건정보를 불러오지 못하고 있는게 원인이지 않나 추측하고 있습니다.

임시대책으로서 `EnsureConditionLoaded()` 대신 `LoadCondition()` 함수를 사용해
직접 강제로 조건정보를 최초 1회 불러오도록 한 뒤에 사용해보시기 바랍니다.

---

구체적으로 설명드려보자면, 제가 `EnsureConditionLoaded()` 함수 최초 구현시 현재 조건정보가 로딩되어 있는지에 대한 판단을 아래와 같은 [공식문서의 `GetConditionLoad()` 함수 설명](https://download.kiwoom.com/web/openapi/kiwoom_openapi_plus_devguide_ver_1.5.pdf#page=26)을 참고하여:

> System 폴더에 아이디_NewSaveIndex.dat파일로 저장된다. Ocx가 종료되면 삭제시킨다.

**조건정보 임시파일의 존재여부 == 조건정보 로딩 여부** 로 동일시해 판단하도록 했고,
따라서 `EnsureConditionLoaded()` 함수의 경우 해당 임시파일이 없을때만 실제 조건정보를 로딩하도록 되어있는데요.

몇번 자체적으로 테스트를 해보니, 해당 파일이 제대로 삭제가 안되서 그런지는 모르겠지만, 임시파일의 존재여부와는 상관없이 무조건 최초 OCX 프로그램 구동 이후에 조건정보를 한번 이상 불러와야만 정상적으로 조건검색을 사용할 수 있는 듯 합니다. 사실 해당 내용이 문서에도 바로 밑에 있긴 하네요...

> 조건검색 사용시 이함수를 최소 한번은 호출해야 조건검색을 할 수 있다.

혹시나 저 문서의 설명처럼 "OCX 종료시 임시파일이 삭제된다" 는 내용이 거짓인건 아닌지 의심하고 싶지만,
그게 아니라면 KOAPY 에서 뭔가 OCX 의 종료시 동작을 처리할 시간을 충분히 주지 않고 강제종료 시켜버리는건 아닌지 하는 생각도 듭니다. (Exit 관련해서 두분께서 말씀주신 내용을 보다보니...)

**Comment by solverkim** [comment]
context.EnsureConditionLoaded() 대신에
context.LoadCondition() 를 사용하니 정상적으로 조회가 됩니다.

감사합니다.
